### DISCUSSION TITLE
SO 2.3.61 "Push to TheHive" scripted field is triggering fielddata circuit breaker / excess heap usage

### DISCUSSION THREAD

**Discussion by ceeeekay** [Security-Onion-Solutions/securityonion#6549] [Unsupported Versions] [answered]
HI there, we've recently started to experience shard failures on the Home dashboard, which point to the Hive URL scripted field being the culprit, particularly the fact it's referencing the `_id` field in each document.

As the circuit breaker is being hit, it's consuming quite a bit of RAM across the cluster - 40% of our total heap, which is well in excess of 100GB wasted across the cluster each time this script runs, and is triggering a lot of unnecessary garbage collection and poor cluster response.

According to Elasticsearch docs, the `_id` field is not supposed to be used in any aggregations or scripts, and running queries and aggs on it is known to cause fielddata blowouts like this.

https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-id-field.html

`"The _id field is restricted from use in aggregations, sorting, and scripting. In case sorting or aggregating on the _id field is required, it is advised to duplicate the content of the _id field into another field that has doc_values enabled."`

As per the above is there any way (without reindexing) that we can have the content of the `_id` field replicated somewhere else that we can run this scripted field against?

I also note that scripted fields are deprecated in favour of runtime fields, so perhaps this would be a good opportunity to replace this field with something else?

Cheers

**Comment by fluffyunicorn24** [comment]
Sorry for tagging onto your post, but I am also having the same issue and came on here to ask that question.  Is there a way we can change it to use keyword type mappings rather than text type for aggregations and sorting? I was thinking of updating setting a fielddata frequency filter on the mapping to limit the amount of fielddata put into memory, but I am not sure if this will lead to further issues.  

I read through all the other discussions that seemed related and none of those solutions were applicable.  I have also upped the JVM Heap and reduced the open time for the indices this did not work.  I appreciate any advice on this.  Thank you.

**Comment by weslambert** [comment]
> HI there, we've recently started to experience shard failures on the Home dashboard, which point to the Hive URL scripted field being the culprit, particularly the fact it's referencing the `_id` field in each document.
> 
> As the circuit breaker is being hit, it's consuming quite a bit of RAM across the cluster - 40% of our total heap, which is well in excess of 100GB wasted across the cluster each time this script runs, and is triggering a lot of unnecessary garbage collection and poor cluster response.
> 
> According to Elasticsearch docs, the `_id` field is not supposed to be used in any aggregations or scripts, and running queries and aggs on it is known to cause fielddata blowouts like this.
> 
> https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-id-field.html
> 
> `"The _id field is restricted from use in aggregations, sorting, and scripting. In case sorting or aggregating on the _id field is required, it is advised to duplicate the content of the _id field into another field that has doc_values enabled."`
> 
> As per the above is there any way (without reindexing) that we can have the content of the `_id` field replicated somewhere else that we can run this scripted field against?
> 
> I also note that scripted fields are deprecated in favour of runtime fields, so perhaps this would be a good opportunity to replace this field with something else?
> 
> Cheers

I'll have to look in to this.  I agree that a runtime field is probably the best approach, however, we 've not had many folks report this issue before, at least that I could find quickly.  How many days worth of indices do you have open, and how large are the indices?

**Comment by ceeeekay** [comment]
@fluffyunicorn24 from what I can tell it's not advised to modify the mapping type for `_id` as it's a reserved field, and from 8.x is going to be more restricted anyway and will possibly be incompatible with your proposed changes.

@weslambert I was thinking about adding something like this to the Logstash pipeline, as a method of removing duplicates in the case that any events were replayed, and it may just solve this whole problem -- although I'm not sure if `document_id` will work if passed to an Ingest Pipeline:

```
    filter {
      fingerprint {
        source              => [ "[@timestamp]", "[log][offset]", "[host][name]", "[message]" ]
        concatenate_sources => true
        target              => "[event][id]"
      }
    }

[...]

    output {
      if [event][id] {
        elasticsearch {
          hosts              => [ "<es_host>" ]
          index              => "<index_name>"
          document_id        => "%{[event][id]}"
          [...]
        }
      } else { ... }
    }
```

This would mean that `_id` would be the same hash from `[event][fingerprint]`, and the value in `[event][fingerprint]` could be mapped as a keyword and queried without interfering with the `_id` field. It also means that we would have a record of the `_id` field before the event is indexed instead of having to retrieve it somehow after index time if we want to query/agg/script it. Bonus is any replayed events would just update/overwrite the doc at `_id` instead of being indexed as a new doc.

If you would consider adding something like this upstream, that would be preferable to users having to change a fundamental SO method, namely the way Logstash outputs every event to Elasticsearch.

Cheers!

[edit] I've updated this to replace `[event][fingerprint]` with `[event][id]` to align with ECS https://www.elastic.co/guide/en/ecs/current/ecs-event.html#field-event-id

**Comment by wildlyunder** [comment]
Hi all, was this pushed upstream as we're having the same issue.

**Comment by dougburks** [comment, marked as answer]
`Push to TheHive` was removed from the Kibana dashboards in Security Onion 2.3.100:
https://blog.securityonion.net/2022/01/security-onion-23100-now-available.html

**Reply by fluffyunicorn24** [reply]
Thank you Doug!
